### PR TITLE
Add basic MD parser tests

### LINKED ISSUE #54: parser/md: add tests
They were removed in https://github.com/googlecodelabs/tools/pull/33. Add them back.

/cc @samtstern

### PR DESCRIPTION
**Pull request by samtstern**
Fixes #54 

Adds basic tests for markdown parsing, matching the intent of the tests deleted in #33.

### CODE REVIEW

File: claat/parser/md/parse.go [hunk 1]
```diff
@@ -33,9 +33,9 @@ import (
 	"golang.org/x/net/html"
 	"golang.org/x/net/html/atom"
 
-	"github.com/googlecodelabs/tools/claat/parser"
 	"github.com/googlecodelabs/tools/claat/types"
 	"github.com/russross/blackfriday"
+	"github.com/googlecodelabs/tools/claat/parser"
```
Comment (x1ddos) [inline claat/parser/md/parse.go]: This doesn't look right. Feels like gofmt wasn't run on this file?
I mean, the imports seem out of order.

File: claat/parser/md/parse.go [hunk 2]
```diff
@@ -365,10 +365,12 @@ func addMetadataToCodelab(m map[string]string, c *types.Codelab) error {
 		case metaCategories:
 			// Standardize the categories and append to codelab field.
 			c.Categories = append(c.Categories, standardSplit(v)...)
+			sort.Strings(c.Categories)
 			break
 		case metaEnvironments:
 			// Standardize the tags and append to the codelab field.
 			c.Tags = append(c.Tags, standardSplit(v)...)
+			sort.Strings(c.Tags)
```
Comment (x1ddos) [inline claat/parser/md/parse.go]: Isn't sorting already done in `func parseMarkup` at the very end?
https://github.com/googlecodelabs/tools/blob/c0f3afc2f3a7436818b5163d1465f6c07e33e32d/claat/parser/md/parse.go#L213

File: claat/parser/md/parse.go [hunk 3]
```diff
@@ -387,6 +389,7 @@ func addMetadataToCodelab(m map[string]string, c *types.Codelab) error {
 		case metaTags:
 			// Standardize the tags and append to the codelab field.
 			c.Tags = append(c.Tags, standardSplit(v)...)
+			sort.Strings(c.Tags)
```
Comment (x1ddos) [inline claat/parser/md/parse.go]: Same here. I thought it was done at the very end.
https://github.com/googlecodelabs/tools/blob/c0f3afc2f3a7436818b5163d1465f6c07e33e32d/claat/parser/md/parse.go#L213

File: claat/parser/md/parse_test.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	c := mustParseCodelab(fmt.Sprintf("# %s", title))
+
+	if c.Title != title {
+		t.Errorf("[%q] got %v, want %v", title, c.Title, title)
+	}
+}
+
+// The parser assumes
+//   * Any single number is minutes
+//   * xx::yy is mm::ss
```
Comment (x1ddos) [inline claat/parser/md/parse_test.go]: why double colon?

File: claat/parser/md/parse_test.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		// Test more than two digits in the hours place.
+		{"100:00:00", 100 * time.Hour},
+		// Test an empty string.
+		{"", 0},
+		// Test complete nonsense.
+		{"Complete nonsense.", 0},
+	}
+
+	for i, tc := range tests {
+		content := fmt.Sprintf(tmp, tc.in)
```
Comment (x1ddos) [inline claat/parser/md/parse_test.go]: I would inline `tmp` here. It makes reading the test easier.

File: claat/parser/md/parse_test.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+		c := mustParseCodelab(content)
+		out := time.Duration(c.Duration) * time.Minute
```
Comment (x1ddos) [inline claat/parser/md/parse_test.go]: Damn, I didn't realize `c.Duration` is not `time.Duration`. We should fix it.
Created https://github.com/googlecodelabs/tools/issues/59.

File: claat/parser/md/parse_test.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+	for i, tc := range tests {
+		content := stdHeader
+		for _, dur := range tc.in {
+			content += fmt.Sprintf(tmp, dur)
+		}
+
+		c := mustParseCodelab(content)
+		if c.Duration != tc.out {
+			t.Errorf("%d: Wanted duration %d but got %d", i, tc.out, c.Duration)
```
Comment (x1ddos) [inline claat/parser/md/parse_test.go]: Reverse the args to match the `if`; lower case:

```go
t.Errorf("%d: duration is %d; want %d", i, c.Duration, tc.out)
```

File: claat/parser/md/parse_test.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+---
+`
+	content += ("# " + title)
+
+	c := mustParseCodelab(content)
+	if !reflect.DeepEqual(out, c.Meta) {
+		t.Errorf("\ngot:\n%+v\nwant:\n%+v", c.Meta, out)
+	}
+}
```
Comment (x1ddos) [inline claat/parser/md/parse_test.go]: No new line in the end?

File: claat/parser/md/parse_test.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+environments: kiosk, web
+analytics account: 12345
+feedback link: https://www.google.com
+
+---
+`
+	content += ("# " + title)
+
+	c := mustParseCodelab(content)
+	if !reflect.DeepEqual(out, c.Meta) {
```
Comment (x1ddos) [inline claat/parser/md/parse_test.go]: The `out` is usually what gets returned from a function under the test. It's unusually to see `out` as a correct value you're testing the actual *output* of a function.

I would name it `wantMeta`, `testMeta`, just `meta` or something like that.

Comment (x1ddos) [inline claat/parser/md/parse_test.go]: Reverse the order of `c.Meta` and `out`:
https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

File: claat/parser/md/parse.go (hunk 2 above)
Comment (samtstern) [reply claat/parser/md/parse.go]: Ah you're right.  What I thought was unstable ordering was actually not!

**Comment by samtstern** [comment]
@x1ddos thanks for the review!  All comments addressed.

**Comment by x1ddos** [comment]
Thanks Sam. I'll leave it to you whenever you want to merge.

**Pull request merged**